### PR TITLE
[MRG] Add SPA method

### PR DESCRIPTION
**Pull request by tgnassou**
https://arxiv.org/pdf/2310.17594

### CODE REVIEW

File: skada/deep/losses.py [hunk 1]
```diff
@@ -414,3 +408,122 @@ def mcc_loss(y, T=1, eps=1e-7):
...
+    memory_outputs : torch.Tensor
+        Memory outputs.
+    K : int, default=5
+        The number of nearest neighbors.
+    sample_idx : torch.Tensor, default=None
+        The sample indices in the batch.
+    """
+    dis = -torch.mm(features_t.detach(), memory_features.t())
+    for i in range(dis.size(0)):
+        dis[i, sample_idx[i]] = torch.inf
```
Comment (YanisLalou) [inline skada/deep/losses.py:520]: Can't we do instead: `dis.scatter_(1, sample_idx, torch.inf)`

File: skada/deep/_graph_alignment.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        )
+        domain_label_target = torch.ones(
+            (domain_pred_t.size()[0]),
+            device=domain_pred_t.device,
+        )
+
+        # update classification function
+        loss_adv = self.domain_criterion_(
+            domain_pred_s, domain_label
+        ) + self.domain_criterion_(domain_pred_t, domain_label_target)
```
Comment (antoinecollas) [inline skada/deep/_graph_alignment.py:92]: in eq (1) of the paper, they use both the cross entropy and the modified cross entropy loss (log(1-prob))

Comment (antoinecollas) [reply skada/deep/_graph_alignment.py:92]: we should also use label smoothing as the code of the paper: https://github.com/CrownX/SPA/blob/main/code/loss.py#L65-L84

File: skada/_mapping.py [hunk 3]
```diff
@@ -97,9 +97,11 @@ def transform(
...
         if X_source.shape[0] > 0:
             X_source = self.ot_transport_.transform(Xs=X_source)
-        X_adapt, _ = source_target_merge(
-            X_source, X_target, sample_domain=sample_domain
-        )
+            X_adapt, _ = source_target_merge(
+                X_source, X_target, sample_domain=sample_domain
+            )
+        else:
+            X_adapt = X_target
```
Comment (antoinecollas) [inline skada/_mapping.py:104]: I think the [#280](https://github.com/scikit-adaptation/skada/pull/280) solved this problem, right?

File: skada/deep/_graph_alignment.py [hunk 4]
```diff
@@ -53,6 +53,7 @@ def __init__(
         reg_adv=1,
         reg_gsa=1,
         reg_nap=1,
+        max_iter=100,
```
Comment (antoinecollas) [inline skada/deep/_graph_alignment.py:56]: max_epochs?

File: skada/deep/_graph_alignment.py [hunk 5]
```diff
@@ -87,19 +91,28 @@ def forward(
         )
 
         # update classification function
-        loss_adv = self.domain_criterion_(
-            domain_pred_s, domain_label
-        ) + self.domain_criterion_(domain_pred_t, domain_label_target)
+        loss_adv = (
+            self.reg_adv
+            * eff
```
Comment (antoinecollas) [inline skada/deep/_graph_alignment.py:96]: We did not put a scheduler in other adversarial methods. I think you should remove it to be consistent.

File: skada/deep/_graph_alignment.py [hunk 6]
```diff
@@ -66,6 +67,8 @@ def __init__(
         self.K = K
         self.memory_features = memory_features
         self.memory_outputs = memory_outputs
+        self.max_iter = max_iter
+        self.n_iter = 0
```
Comment (antoinecollas) [inline skada/deep/_graph_alignment.py:71]: n_epoch?

File: skada/_mapping.py [hunk 7]
```diff
@@ -95,8 +95,6 @@ def transform(
         X_source, X_target = source_target_split(X, sample_domain=sample_domain)
         # in case of prediction we would get only target samples here,
         # thus there's no need to perform any transformations
-        if X_source.shape[0] > 0:
-            X_source = self.ot_transport_.transform(Xs=X_source)
```
Comment (antoinecollas) [inline skada/_mapping.py:99]: you want to remove these two lines ?

File: skada/deep/losses.py [hunk 8]
```diff
@@ -414,3 +408,109 @@ def mcc_loss(y, T=1, eps=1e-7):
...
+    memory_outputs : torch.Tensor
+        Memory outputs.
+    sample_idx : torch.Tensor
+        The sample indices in the batch.
+    K : int, default=5
+        The number of nearest neighbors.
+    """
+    dis = -torch.mm(features_t.detach(), memory_features.t())
+
+    dis[torch.arange(dis.size(0)), sample_idx] = float("-inf")
```
Comment (antoinecollas) [inline skada/deep/losses.py:503]: it should be +inf instead of -inf?

File: skada/deep/losses.py [hunk 9]
```diff
@@ -414,3 +408,109 @@ def mcc_loss(y, T=1, eps=1e-7):
...
+
+    # Get top-K neighbors
+    _, top_k_indices = torch.topk(dis, k=K, dim=1)
```
Comment (antoinecollas) [inline skada/deep/losses.py:506]: it should be `torch.topk(-dis, k=K, dim=1)`, right?

File: skada/deep/losses.py (hunk 8 above)
Comment (YanisLalou) [reply skada/deep/losses.py:503]: I agree

File: skada/deep/callbacks.py [hunk 10]
```diff
@@ -73,3 +74,104 @@ def on_epoch_begin(self, net, dataset_train=None, **kwargs):
             target_kmeans.fit(features_t)
 
         net.criterion__adapt_criterion.target_kmeans = target_kmeans
+
+
+class ComputeMemoryBank(Callback):
```
Comment (antoinecollas) [inline skada/deep/callbacks.py:79]: Minor remark: I think `ComputeMemoryBank` and `MemoryBankInit` should be merged. Wdyt?

**Comment by YanisLalou** [comment]
Nice results with SPA, lets merge this now!
